### ISSUE TITLE
Binary Gems for Mac and Linux

### ISSUE THREAD

**Issue by AndyObtiva** [mcorino/wxRuby3#229]
Thank you for continuing to maintain the wxRuby3 binding.

Now that the dust settled from my [RubyConf 2023 2-Hour Workshop "How To Build Desktop Applications in Ruby"](https://github.com/AndyObtiva/how-to-build-desktop-applications-in-ruby), I can start focusing on libraries other than [Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui), which was used in the workshop. One of the biggest weaknesses of the LibUI toolkit is that it is immature as a mid-alpha, thus missing several features for desktop GUI development like support for Tree widget, Image widget, and App Window Icon property. One of the biggest strengths of Glimmer DSL for LibUI is it includes binaries for all platforms of Mac, Windows, and Linux, so people can install the Ruby gem quickly and get started instantly. 

If you were to package binaries for Mac and Linux in the wxRuby3 gem, that would make the value proposition of using the mature wxWidgets better in Ruby, removing the trade-off it has with LibUI as far as gem installation time/convenience, and I could continue to add value on top of that with a declarative minimal-syntax highly-productive Ruby DSL in [Glimmer DSL for WX](https://github.com/AndyObtiva/glimmer-dsl-wx).

You can perhaps support binary gems for Mac and Linux in an incremental multi-release approach.

**Comment by mcorino** [comment]
I'll probably start looking into this around the time I'm planning to (or after) release the 1.0.0 version.

Unfortunately this is not going to be as easy as it is for LibUI as wxWidgets not only pulls in more dependencies but is also a C++ library instead of a C-library like LibUI (which means even more platform dependencies).

The Windows binary package was relatively easy as 1. it pulls in less dependencies than other platforms since it relies more on standard Windows libraries and 2. the C++ dependencies are actually part of the Ruby package on that platform.

Overall I see three options:

1. make the source build 'smarter' as in 'have it pull in (more of) it's dependencies automatically depending on the platform'; this might make the installation (almost) as easy as a prebuilt binary (apart from the waiting time);
2. use a service like OBS to provide prebuilt distro packages;
3. provide (more) binary gems.

For OSX option 3 might be the best solution. For Linux I'm currently leaning towards a combination of 1 and 2.

However this is not my highest priority at the moment. I still have a number of technical features I want to add first.

**Comment by rubyFeedback** [comment]
Strictly speaking while Linux binaries may be convenient, it should be trivial for Linux users to first install the wxwidgets-relevant libraries (e. g. via apt-get etc...), and then - if they have a sane toolchain for compilation - just compile the wxruby3 bindings from source. I have seen other projects that went nuts and offered binaries for about the six most commonly used linux distributions, and I always felt this was not the best way to go about it, since that encouraged fragmentation (each distribution doing their own thing), along with an increase in time investment from the maintainer(s). So I think for wxRuby3, it may be best to document recommended practices on different platforms - including linux - and let linux users choose their preferred method; this could be documented on the wiki too, if it would inflate the main README too much; the jruby folks, for instance, document some things on their FAQ entry which resides on the github wiki.

My preferred way to install things on linux is to simply compile from source, for instance, using ruby to aid in the process. If the toolchain is sane then this is actually a very trivial way to maintain a linux system; I only opt for alternatives for large projects, such as libreoffice, where I simply install the .deb package (I use manjaro usually these days, so I also use ruby to extract the .deb archive and then repackage it into a .tar.xz so I can install it on other computer systems too).

TL;DR: I think whichever way you decide to go for, it is best to describe it, so that other users can read up on it and decide how they want to go about it. For experienced users this is usually not so important, but new users or semi-new users may be helped that way.

**Comment by mcorino** [comment]
@AndyObtiva The new v0.9.8 release introduces support for and release of binary packages. This should bring you more or less the prerequisite free installation you like. 

Remaining "problem" is that, as I expected, it is impossible to create cross-distibution/-version packages for Linux because of the extensive wxWidgets dependencies so for Linux each version of each distro would require a separate binary package for which there simply is not enough **FREE** CPU time to be had from the various CI/CD suppliers.
In addition Ruby MRI binary packages need to be Ruby ABI specific making the number of possible distinct packages even bigger.
For now I opted only to provide packages for the latest versions of each major distro supported and one each for the standard system supplied Ruby and for the latest stable Ruby MRI.
For MacOSX I also opted for OSX major version specific packages (instead of Universal packages).
And as Windows  and MacOSX do not provide (usable) system provided Rubies for these systems I only provide packages for the latest stable Ruby MRI.

Have fun.

**Comment by AndyObtiva** [comment]
Thank you very much!
